runtests.py

```diff
@@ -819,6 +827,10 @@ def build_tests(self, test_class, path, workdir, module, module_path, expect_err
         if 'cpp' in languages and 'no-cpp' in tags['tag']:
             languages = list(languages)
             languages.remove('cpp')
+        if (self.add_cpp_locals_extra_tests and 'cpp' in languages and
+                'cpp' in tags['tag'] and not 'no-cpp-locals' in tags['tag']):
+            languages = list(languages)
+            languages.append('cpp_locals')
```

Comment:
Hmm. A 'language' doesn't really seem the right way to express this. We already split tests for `.py` files. Are the requirements for this tag really different?

Reply:
I agree it's not quite a language. The nice thing about misusing language is that it creates them in a separate cpp_locals directory and reports the tests as cpp_locals/testnames without too much extra effort.

The approach we use for the .py files doesn't quite work - that creates a separate `PureDoctestTestCase` at a much earlier stage before the `CythonRunTestCase` is created (which makes sense there because it's quite a different test, but doesn't make so much sense here where it's largely the same test but with different options).

What would work is changing:

```
tests = [ self.build_test(test_class, path, workdir, module, module_path,
                                  tags, language, language_level,
                                  expect_errors, expect_warnings, warning_errors, preparse,
                                  pythran_dir if language == "cpp" else None,
                                  add_cython_import=add_cython_import)
                  for language in languages
                  for preparse in preparse_list
                  for language_level in language_levels
        ]
```
to
```
tests = [ self.build_test(test_class, path, workdir, module, module_path,
                                  tags, language, language_level,
                                  expect_errors, expect_warnings, warning_errors, preparse,
                                  pythran_dir if language == "cpp" else None,
                                  add_cython_import=add_cython_import,
                                  cpp_locals_on=cpp_locals_on)
                  for language in languages
                  for preparse in preparse_list
                  for language_level in language_levels
                  for cpp_locals_on in cpp_locals_states
        ]
```

(plus a few supporting changes elsewhere). To me that feels more intrusive that using 'language', but I'm happy to make that change if you disagree.

Reply:
This is pretty much what needs to be done, I think, but we should use a list of `extra_directives` dicts. Given that we will eventually need to test at least some directives separately or in combination, I think this one isn't special enough to merit its own option.

You can keep the `no_cpp_locals` tag by simply checking if the part without the `no_` is in the extra directives.

Reply:
Yeah a generic solution is probably better so I've done that. In principle it might be possible to handle language level in the extra directives too but I decided it was safer not to touch that for now.